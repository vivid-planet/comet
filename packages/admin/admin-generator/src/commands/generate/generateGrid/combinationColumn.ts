import { type GridColDef } from "@comet/admin";
import { type ComponentProps } from "react";
import { type FormattedNumber } from "react-intl";

import { type BaseColumnConfig } from "../generate-command";
import { getFormattedMessageNode } from "../utils/intl";

type AbstractField<FieldName extends string> = {
    field: FieldName;
    emptyValue?: string;
};

type TextField<FieldName extends string> = AbstractField<FieldName> & {
    type: "text";
};

type StaticText = {
    type: "static";
    text: string;
};

type FormattedNumberPropsForNumberField = Omit<ComponentProps<typeof FormattedNumber>, "value" | "children">;

type NumberField<FieldName extends string> = AbstractField<FieldName> &
    FormattedNumberPropsForNumberField & {
        type: "number";
        decimals?: number;
    };

type StaticSelectField<FieldName extends string> = AbstractField<FieldName> & {
    type: "staticSelect";
    values: Array<
        | string
        | {
              value: string | number | boolean;
              label: string;
          }
    >;
};

type FormattedMessage<FieldName extends string> = {
    type: "formattedMessage";
    message: string;
    valueFields: Record<string, Field<FieldName>>;
};

type Field<FieldName extends string> =
    | StaticText
    | FieldName
    | TextField<FieldName>
    | NumberField<FieldName>
    | StaticSelectField<FieldName>
    | FormattedMessage<FieldName>;

export type GridCombinationColumnConfig<FieldName extends string> = {
    type: "combination";
    name: string;
    primaryText?: Field<FieldName>;
    secondaryText?: Field<FieldName>;
} & BaseColumnConfig &
    Pick<GridColDef, "sortBy">;

type CellContent = {
    textContent: string;
    variableDefinitions?: string[];
};

const getTextForCellContent = (textConfig: Field<string>, messageIdPrefix: string): CellContent => {
    if (typeof textConfig !== "string" && textConfig.type === "static") {
        return {
            textContent: getFormattedMessageNode(messageIdPrefix, textConfig.text),
        };
    }

    if (typeof textConfig !== "string" && textConfig.type === "formattedMessage") {
        const variableDefinitions: string[] = [];

        const values = Object.entries(textConfig.valueFields)
            .map(([key, value]) => {
                const { textContent, variableDefinitions: cellVariableDefinitions } = getTextForCellContent(value, `${messageIdPrefix}.${key}`);

                if (cellVariableDefinitions?.length) {
                    variableDefinitions.push(...cellVariableDefinitions);
                }

                return `${key}: ${textContent}`;
            })
            .join(", ");

        return {
            textContent: getFormattedMessageNode(messageIdPrefix, textConfig.message, `{${values}}`),
            variableDefinitions,
        };
    }

    const emptyText =
        typeof textConfig !== "string" && "emptyValue" in textConfig
            ? getFormattedMessageNode(`${messageIdPrefix}.empty`, textConfig.emptyValue)
            : "'-'";

    const fieldName = typeof textConfig === "string" ? textConfig : textConfig.field;
    const rowValue = `row.${fieldName.replace(/\./g, "?.")}`;
    const stringTextContent = `${rowValue} ?? ${emptyText}`;

    if (typeof textConfig === "string") {
        return {
            textContent: stringTextContent,
        };
    }

    if (textConfig.type === "number") {
        const { type, decimals: decimalsConfigValue, field, emptyValue, ...configForFormattedNumberProps } = textConfig;

        const hasCurrency = Boolean(textConfig.currency);
        const hasUnit = Boolean(textConfig.unit);

        const defaultDecimalsProp = hasCurrency && decimalsConfigValue === undefined ? 2 : decimalsConfigValue;
        let defaultStyleProp: string | undefined = undefined;

        if (hasCurrency) {
            defaultStyleProp = '"currency"';
        } else if (hasUnit) {
            defaultStyleProp = '"unit"';
        }

        const formattedNumberProps: Record<string, unknown> = {
            value: `{${rowValue}}`,
            minimumFractionDigits: typeof defaultDecimalsProp !== "undefined" ? `{${defaultDecimalsProp}}` : undefined,
            maximumFractionDigits: typeof defaultDecimalsProp !== "undefined" ? `{${defaultDecimalsProp}}` : undefined,
            style: typeof defaultStyleProp !== "undefined" ? defaultStyleProp : undefined,
        };

        Object.entries(configForFormattedNumberProps).forEach(([key, value]) => {
            if (typeof value === "string") {
                formattedNumberProps[key] = `"${value}"`;
            } else {
                formattedNumberProps[key] = `{${value}}`;
            }
        });

        const formattedNumberPropsString = Object.entries(formattedNumberProps)
            .map(([key, value]) => {
                if (typeof value === "undefined") {
                    return null;
                }
                return `${key}=${value}`;
            })
            .join(" ");

        return {
            textContent: `typeof ${rowValue} === "undefined" || ${rowValue} === null ? ${emptyText} : <FormattedNumber ${formattedNumberPropsString} />`,
        };
    }

    if (textConfig.type === "staticSelect") {
        const labelsVariableName = `${textConfig.field}Labels`;

        const labelMapping = textConfig.values
            .map((valueOption) => {
                const value = typeof valueOption === "string" ? valueOption : valueOption.value;
                const label = typeof valueOption === "string" ? valueOption : valueOption.label;
                return `${value}: ${getFormattedMessageNode(`${messageIdPrefix}.${value}`, label)}`;
            })
            .join(", ");

        const labelMappingVar = `const ${labelsVariableName}: Record<string, ReactNode> = { ${labelMapping} };`;
        const textContent = `(${rowValue} == null ? ${emptyText} : ${labelsVariableName}[` + `\`\${${rowValue}}\`` + `] ?? ${rowValue})`;

        return {
            textContent,
            variableDefinitions: [labelMappingVar],
        };
    }

    return {
        textContent: stringTextContent,
    };
};

export const getCombinationColumnRenderCell = (column: GridCombinationColumnConfig<string>, messageIdPrefix: string) => {
    const gridCellContentProps: Record<string, string> = {};
    const allVariableDefinitions: string[] = [];

    if (column.primaryText) {
        const { textContent, variableDefinitions = [] } = getTextForCellContent(column.primaryText, `${messageIdPrefix}.primaryText`);
        gridCellContentProps.primaryText = textContent;
        allVariableDefinitions.push(...variableDefinitions);
    }

    if (column.secondaryText) {
        const { textContent, variableDefinitions = [] } = getTextForCellContent(column.secondaryText, `${messageIdPrefix}.secondaryText`);
        gridCellContentProps.secondaryText = textContent;
        allVariableDefinitions.push(...variableDefinitions);
    }

    const allUniqueVariableDefinitions = Array.from(new Set(allVariableDefinitions));

    return `({ row }) => {
        ${allUniqueVariableDefinitions.join("\n")}
        return <GridCellContent ${Object.entries(gridCellContentProps)
            .map(([key, value]) => `${key}={${value}}`)
            .join(" ")} />;
    }`;
};

const getFieldNamesFromText = (textConfig: Field<string>): string[] => {
    if (typeof textConfig === "string") {
        return [textConfig];
    }

    if (textConfig.type === "static") {
        return [];
    }

    if (textConfig.type === "formattedMessage") {
        return Object.values(textConfig.valueFields).flatMap((value) => getFieldNamesFromText(value));
    }

    return [textConfig.field];
};

export const getAllColumnFieldNames = (column: GridCombinationColumnConfig<string>): string[] => {
    const fieldNames: string[] = [];

    if (column.primaryText) {
        fieldNames.push(...getFieldNamesFromText(column.primaryText));
    }

    if (column.secondaryText) {
        fieldNames.push(...getFieldNamesFromText(column.secondaryText));
    }

    return fieldNames.filter((fieldName) => fieldName !== "id");
};
