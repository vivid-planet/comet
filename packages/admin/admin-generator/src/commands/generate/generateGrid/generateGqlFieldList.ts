import objectPath from "object-path";

import { type ActionsGridColumnConfig, type GridColumnConfig } from "../generate-command";
import { getAllColumnFieldNames, type GridCombinationColumnConfig } from "./combinationColumn";

type FieldsObjectType = { [key: string]: FieldsObjectType | boolean | string };
const recursiveStringify = (obj: FieldsObjectType): string => {
    let ret = "";
    let prefixField = "";
    for (const key in obj) {
        const valueForKey = obj[key];
        if (typeof valueForKey === "boolean") {
            ret += `${prefixField}${key}`;
        } else if (typeof valueForKey === "string") {
            ret += `${prefixField}${key}${valueForKey}`;
        } else {
            ret += `${prefixField}${key} { ${recursiveStringify(valueForKey)} }`;
        }
        prefixField = " ";
    }
    return ret;
};

export function generateGqlFieldList({
    columns,
}: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    columns: Array<GridColumnConfig<any> | GridCombinationColumnConfig<string> | ActionsGridColumnConfig>;
}) {
    const fieldsObject: FieldsObjectType = columns.reduce<FieldsObjectType>((acc, field) => {
        if (field.type !== "actions") {
            if (field.type === "combination") {
                getAllColumnFieldNames(field).map((fieldName) => {
                    objectPath.set(acc, fieldName, true);
                });
            } else {
                objectPath.set(acc, field.name, true);
            }
        }
        return acc;
    }, {});
    return recursiveStringify(fieldsObject);
}
