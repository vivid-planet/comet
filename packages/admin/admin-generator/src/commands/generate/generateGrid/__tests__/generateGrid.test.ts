import { buildSchema, type GraphQLSchema, introspectionFromSchema, type IntrospectionQuery } from "graphql";

import { type GridConfig } from "../../generate-command";
import { generateGrid } from "../generateGrid";

describe("generateGrid", () => {
    let schema: GraphQLSchema;
    let introspection: IntrospectionQuery;

    beforeAll(() => {
        schema = buildSchema(`
            input StringFilter {
                equal: String
            }

            type Query {
                books(
                    offset: Int!
                    limit: Int!
                    sort: [BookSort!]
                    filter: BookFilter
                ): PaginatedBooks!
                booksByAuthor(
                    authorId: ID!
                    offset: Int!
                    limit: Int!
                    sort: [BookSort!]
                    filter: BookFilter
                ): PaginatedBooks!
            } 

            type PaginatedBooks {
                nodes: [Book!]!
                totalCount: Int!
            }

            type Book {
                id: ID!
                title: String!
            }

            input BookFilter {
                title: StringFilter
            }

            input BookSort {
                field: BookSortField!
                direction: SortDirection!
            }

            enum BookSortField {
                title
            }

            enum SortDirection {
                ASC
                DESC
            }

            type Mutation {
                createBook(input: BookInput!): Book!
                updateBook(id: ID!, input: BookInput!): Book!
                deleteBook(id: ID!): Boolean!
            }

            input BookInput {
                title: String!
            }
        `);

        introspection = introspectionFromSchema(schema);
    });

    type Book = {
        __typename: "Book";
        id: string;
        title: string;
    };

    it("should generate a grid with book configuration", () => {
        const config: GridConfig<Book> = {
            type: "grid",
            gqlType: "Book",
            columns: [
                {
                    type: "text",
                    name: "title",
                },
            ],
        };

        const result = generateGrid(
            {
                exportName: "BooksGrid",
                baseOutputFilename: "BooksGrid",
                targetDirectory: "/test",
                gqlIntrospection: introspection,
            },
            config,
        );

        expect(result.code).toMatchSnapshot();
    });

    it("should generate required root gql args in export-query variables", () => {
        const config: GridConfig<Book> = {
            type: "grid",
            gqlType: "Book",
            query: "booksByAuthor",
            excelExport: true,
            columns: [
                {
                    type: "text",
                    name: "title",
                },
            ],
        };

        const result = generateGrid(
            {
                exportName: "BooksGrid",
                baseOutputFilename: "BooksGrid",
                targetDirectory: "/test",
                gqlIntrospection: introspection,
            },
            config,
        );

        expect(result.code).toMatchSnapshot();
    });
});
