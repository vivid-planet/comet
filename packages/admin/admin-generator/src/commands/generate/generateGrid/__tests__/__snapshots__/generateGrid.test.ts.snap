// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateGrid should generate a grid with book configuration 1`] = `
"import {
        GQLBooksGridQuery,
        GQLBooksGridQueryVariables,
        GQLBooksFormFragment,
        GQLCreateBookMutation,
        GQLCreateBookMutationVariables,
        GQLUpdateBookPositionMutation,
        GQLUpdateBookPositionMutationVariables,
        GQLDeleteBookMutation,
        GQLDeleteBookMutationVariables
    } from "./BooksGrid.generated";
    import { FormattedMessage } from "react-intl";
import { FormattedNumber } from "react-intl";
import { useIntl } from "react-intl";
import { ReactNode } from "react";
import { gql } from "@apollo/client";
import { useApolloClient } from "@apollo/client";
import { useQuery } from "@apollo/client";
import { Button } from "@comet/admin";
import { CrudContextMenu } from "@comet/admin";
import { CrudMoreActionsMenu } from "@comet/admin";
import { DataGridToolbar } from "@comet/admin";
import { ExportApi } from "@comet/admin";
import { filterByFragment } from "@comet/admin";
import { GridFilterButton } from "@comet/admin";
import { GridCellContent } from "@comet/admin";
import { GridColDef } from "@comet/admin";
import { dataGridDateTimeColumn } from "@comet/admin";
import { dataGridDateColumn } from "@comet/admin";
import { dataGridIdColumn } from "@comet/admin";
import { dataGridManyToManyColumn } from "@comet/admin";
import { dataGridOneToManyColumn } from "@comet/admin";
import { renderStaticSelectCell } from "@comet/admin";
import { messages } from "@comet/admin";
import { muiGridFilterToGql } from "@comet/admin";
import { muiGridSortToGql } from "@comet/admin";
import { StackLink } from "@comet/admin";
import { FillSpace } from "@comet/admin";
import { Tooltip } from "@comet/admin";
import { useBufferedRowCount } from "@comet/admin";
import { useDataGridExcelExport } from "@comet/admin";
import { useDataGridRemote } from "@comet/admin";
import { usePersistentColumnState } from "@comet/admin";
import { BlockPreviewContent } from "@comet/cms-admin";
import { useContentScope } from "@comet/cms-admin";
import { Alert } from "@mui/material";
import { Box } from "@mui/material";
import { IconButton } from "@mui/material";
import { Typography } from "@mui/material";
import { useTheme } from "@mui/material";
import { Menu } from "@mui/material";
import { MenuItem } from "@mui/material";
import { ListItemIcon } from "@mui/material";
import { ListItemText } from "@mui/material";
import { CircularProgress } from "@mui/material";
import { DataGridPro } from "@mui/x-data-grid-pro";
import { GridLogicOperator } from "@mui/x-data-grid-pro";
import { GridRenderCellParams } from "@mui/x-data-grid-pro";
import { GridSlotsComponent } from "@mui/x-data-grid-pro";
import { GridToolbarProps } from "@mui/x-data-grid-pro";
import { GridColumnHeaderTitle } from "@mui/x-data-grid-pro";
import { GridToolbarQuickFilter } from "@mui/x-data-grid-pro";
import { GridRowOrderChangeParams } from "@mui/x-data-grid-pro";
import { Add as AddIcon } from "@comet/admin-icons";
import { Edit as EditIcon } from "@comet/admin-icons";
import { Info as InfoIcon } from "@comet/admin-icons";
import { Excel as ExcelIcon } from "@comet/admin-icons";

    const booksFragment = gql\`
        fragment BooksForm on Book {
            id
            title
        }
    \`;

    const booksQuery = gql\`
        query BooksGrid($offset: Int!, $limit: Int!, $sort: [BookSort!], $filter: BookFilter) {
    books(offset: $offset, limit: $limit, sort: $sort, filter: $filter) {
                nodes {
                    ...BooksForm
                }
                totalCount
            }
        }
        \${booksFragment}
    \`;

    

    const deleteBookMutation = gql\`
                mutation DeleteBook($id: ID!) {
                    deleteBook(id: $id)
                }
            \`;


    
    function BooksGridToolbar() {
        return (
            <DataGridToolbar>
                <GridFilterButton />
                <FillSpace />
        <Button responsive startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add">
        <FormattedMessage id="book.booksForm.newEntry" defaultMessage={\`New Book\`}  />
    </Button>
            </DataGridToolbar>
        );
    }

    

    export function BooksGrid() {
        const client = useApolloClient();
        const intl = useIntl();
        const dataGridProps = { ...useDataGridRemote(), ...usePersistentColumnState("BooksGrid") };
        
        

        

        const columns: GridColDef<GQLBooksFormFragment>[] = [
            {field: "title",
headerName: intl.formatMessage({ id: "book.title", defaultMessage: "Title" }),
flex: 1,
minWidth: 150,},
                {field: "actions",
headerName: "",
sortable: false,
filterable: false,
type: "actions",
align: "right",
pinned: "right",
width: 84,
renderCell: (params) => {
                            return (
                                <>
                                
                                        <IconButton color="primary" component={StackLink} pageName="edit" payload={params.row.id}>
                                            <EditIcon />
                                        </IconButton>
                                        <CrudContextMenu
                                            
                                            onDelete={async () => {
                                                await client.mutate<GQLDeleteBookMutation, GQLDeleteBookMutationVariables>({
                                                    mutation: deleteBookMutation,
                                                    variables: { id: params.row.id },
                                                });
                                            }}
                                            
                                            refetchQueries={[booksQuery]}
                                        />
                                    
                                </>
                            );
                                },}
        ];

        const { filter: gqlFilter,  } = muiGridFilterToGql(columns, dataGridProps.filterModel);

        const { data, loading, error } = useQuery<GQLBooksGridQuery, GQLBooksGridQueryVariables>(booksQuery, {
            variables: {
                filter: gqlFilter, offset: dataGridProps.paginationModel.page * dataGridProps.paginationModel.pageSize, limit: dataGridProps.paginationModel.pageSize, sort: muiGridSortToGql(dataGridProps.sortModel, columns)
            },
        });
        const rowCount = useBufferedRowCount(data?.books.totalCount);
        if (error) throw error;
        const rows = data?.books.nodes ?? [];

        

        return (
            <DataGridPro
                {...dataGridProps}
                rows={rows}
                rowCount={rowCount}
                columns={columns}
                loading={loading}
                slots={{
                                toolbar: BooksGridToolbar as GridSlotsComponent["toolbar"],
                            }}
                            
                
            />
        );
    }
    "
`;

exports[`generateGrid should generate required root gql args in export-query variables 1`] = `
"import {
        GQLBooksGridQuery,
        GQLBooksGridQueryVariables,
        GQLBooksFormFragment,
        GQLCreateBookMutation,
        GQLCreateBookMutationVariables,
        GQLUpdateBookPositionMutation,
        GQLUpdateBookPositionMutationVariables,
        GQLDeleteBookMutation,
        GQLDeleteBookMutationVariables
    } from "./BooksGrid.generated";
    import { FormattedMessage } from "react-intl";
import { FormattedNumber } from "react-intl";
import { useIntl } from "react-intl";
import { ReactNode } from "react";
import { gql } from "@apollo/client";
import { useApolloClient } from "@apollo/client";
import { useQuery } from "@apollo/client";
import { Button } from "@comet/admin";
import { CrudContextMenu } from "@comet/admin";
import { CrudMoreActionsMenu } from "@comet/admin";
import { DataGridToolbar } from "@comet/admin";
import { ExportApi } from "@comet/admin";
import { filterByFragment } from "@comet/admin";
import { GridFilterButton } from "@comet/admin";
import { GridCellContent } from "@comet/admin";
import { GridColDef } from "@comet/admin";
import { dataGridDateTimeColumn } from "@comet/admin";
import { dataGridDateColumn } from "@comet/admin";
import { dataGridIdColumn } from "@comet/admin";
import { dataGridManyToManyColumn } from "@comet/admin";
import { dataGridOneToManyColumn } from "@comet/admin";
import { renderStaticSelectCell } from "@comet/admin";
import { messages } from "@comet/admin";
import { muiGridFilterToGql } from "@comet/admin";
import { muiGridSortToGql } from "@comet/admin";
import { StackLink } from "@comet/admin";
import { FillSpace } from "@comet/admin";
import { Tooltip } from "@comet/admin";
import { useBufferedRowCount } from "@comet/admin";
import { useDataGridExcelExport } from "@comet/admin";
import { useDataGridRemote } from "@comet/admin";
import { usePersistentColumnState } from "@comet/admin";
import { BlockPreviewContent } from "@comet/cms-admin";
import { useContentScope } from "@comet/cms-admin";
import { Alert } from "@mui/material";
import { Box } from "@mui/material";
import { IconButton } from "@mui/material";
import { Typography } from "@mui/material";
import { useTheme } from "@mui/material";
import { Menu } from "@mui/material";
import { MenuItem } from "@mui/material";
import { ListItemIcon } from "@mui/material";
import { ListItemText } from "@mui/material";
import { CircularProgress } from "@mui/material";
import { DataGridPro } from "@mui/x-data-grid-pro";
import { GridLogicOperator } from "@mui/x-data-grid-pro";
import { GridRenderCellParams } from "@mui/x-data-grid-pro";
import { GridSlotsComponent } from "@mui/x-data-grid-pro";
import { GridToolbarProps } from "@mui/x-data-grid-pro";
import { GridColumnHeaderTitle } from "@mui/x-data-grid-pro";
import { GridToolbarQuickFilter } from "@mui/x-data-grid-pro";
import { GridRowOrderChangeParams } from "@mui/x-data-grid-pro";
import { Add as AddIcon } from "@comet/admin-icons";
import { Edit as EditIcon } from "@comet/admin-icons";
import { Info as InfoIcon } from "@comet/admin-icons";
import { Excel as ExcelIcon } from "@comet/admin-icons";

    const booksFragment = gql\`
        fragment BooksForm on Book {
            id
            title
        }
    \`;

    const booksQuery = gql\`
        query BooksGrid($authorId: ID!, $offset: Int!, $limit: Int!, $sort: [BookSort!], $filter: BookFilter) {
    booksByAuthor(authorId: $authorId, offset: $offset, limit: $limit, sort: $sort, filter: $filter) {
                nodes {
                    ...BooksForm
                }
                totalCount
            }
        }
        \${booksFragment}
    \`;

    

    const deleteBookMutation = gql\`
                mutation DeleteBook($id: ID!) {
                    deleteBook(id: $id)
                }
            \`;


    interface BooksGridToolbarToolbarProps extends GridToolbarProps {
         exportApi: ExportApi;
}
    function BooksGridToolbar({
        exportApi
    }: BooksGridToolbarToolbarProps) {
        return (
            <DataGridToolbar>
                <GridFilterButton />
                <FillSpace />
        <CrudMoreActionsMenu
        slotProps={{
            button: {
                responsive: true
            }
        }}
        overallActions={[
            {
                        label: <FormattedMessage {...messages.downloadAsExcel} />,
                        icon: exportApi.loading ? <CircularProgress size={20} /> : <ExcelIcon />,
                        onClick: () => exportApi.exportGrid(),
                        disabled: exportApi.loading,
                    }
        ]}
    />
        <Button responsive startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add">
        <FormattedMessage id="book.booksForm.newEntry" defaultMessage={\`New Book\`}  />
    </Button>
            </DataGridToolbar>
        );
    }

    type Props = {
            authorId: string;
        };

    export function BooksGrid({authorId }: Props) {
        const client = useApolloClient();
        const intl = useIntl();
        const dataGridProps = { ...useDataGridRemote(), ...usePersistentColumnState("BooksGrid") };
        
        

        

        const columns: GridColDef<GQLBooksFormFragment>[] = [
            {field: "title",
headerName: intl.formatMessage({ id: "book.title", defaultMessage: "Title" }),
flex: 1,
minWidth: 150,},
                {field: "actions",
headerName: "",
sortable: false,
filterable: false,
type: "actions",
align: "right",
pinned: "right",
width: 84,
disableExport: true,
renderCell: (params) => {
                            return (
                                <>
                                
                                        <IconButton color="primary" component={StackLink} pageName="edit" payload={params.row.id}>
                                            <EditIcon />
                                        </IconButton>
                                        <CrudContextMenu
                                            
                                            onDelete={async () => {
                                                await client.mutate<GQLDeleteBookMutation, GQLDeleteBookMutationVariables>({
                                                    mutation: deleteBookMutation,
                                                    variables: { id: params.row.id },
                                                });
                                            }}
                                            
                                            refetchQueries={[booksQuery]}
                                        />
                                    
                                </>
                            );
                                },}
        ];

        const { filter: gqlFilter,  } = muiGridFilterToGql(columns, dataGridProps.filterModel);

        const { data, loading, error } = useQuery<GQLBooksGridQuery, GQLBooksGridQueryVariables>(booksQuery, {
            variables: {
                authorId, filter: gqlFilter, offset: dataGridProps.paginationModel.page * dataGridProps.paginationModel.pageSize, limit: dataGridProps.paginationModel.pageSize, sort: muiGridSortToGql(dataGridProps.sortModel, columns)
            },
        });
        const rowCount = useBufferedRowCount(data?.booksByAuthor.totalCount);
        if (error) throw error;
        const rows = data?.booksByAuthor.nodes ?? [];

        const exportApi = useDataGridExcelExport<GQLBooksGridQuery["booksByAuthor"]["nodes"][0], GQLBooksGridQuery, Omit<GQLBooksGridQueryVariables, "offset" | "limit">>({
        columns,
        variables: {
            authorId, ...muiGridFilterToGql(columns, dataGridProps.filterModel)
        },
        query: booksQuery,
        resolveQueryNodes: (data) => data.booksByAuthor.nodes,
        totalCount: data?.booksByAuthor.totalCount ?? 0,
        exportOptions: {
            fileName: "Books",
        },
    });

        return (
            <DataGridPro
                {...dataGridProps}
                rows={rows}
                rowCount={rowCount}
                columns={columns}
                loading={loading}
                slots={{
                                toolbar: BooksGridToolbar as GridSlotsComponent["toolbar"],
                            }}
                            slotProps={{
        toolbar: {                exportApi,    } as BooksGridToolbarToolbarProps  
    }}
                
            />
        );
    }
    "
`;

exports[`generateGrid should generate valueGetter for date in nested field 1`] = `
"import {
        GQLBooksGridQuery,
        GQLBooksGridQueryVariables,
        GQLBooksFormFragment,
        GQLCreateBookMutation,
        GQLCreateBookMutationVariables,
        GQLUpdateBookPositionMutation,
        GQLUpdateBookPositionMutationVariables,
        GQLDeleteBookMutation,
        GQLDeleteBookMutationVariables
    } from "./BooksGrid.generated";
    import { FormattedMessage } from "react-intl";
import { FormattedNumber } from "react-intl";
import { useIntl } from "react-intl";
import { ReactNode } from "react";
import { gql } from "@apollo/client";
import { useApolloClient } from "@apollo/client";
import { useQuery } from "@apollo/client";
import { Button } from "@comet/admin";
import { CrudContextMenu } from "@comet/admin";
import { CrudMoreActionsMenu } from "@comet/admin";
import { DataGridToolbar } from "@comet/admin";
import { ExportApi } from "@comet/admin";
import { filterByFragment } from "@comet/admin";
import { GridFilterButton } from "@comet/admin";
import { GridCellContent } from "@comet/admin";
import { GridColDef } from "@comet/admin";
import { dataGridDateTimeColumn } from "@comet/admin";
import { dataGridDateColumn } from "@comet/admin";
import { dataGridIdColumn } from "@comet/admin";
import { dataGridManyToManyColumn } from "@comet/admin";
import { dataGridOneToManyColumn } from "@comet/admin";
import { renderStaticSelectCell } from "@comet/admin";
import { messages } from "@comet/admin";
import { muiGridFilterToGql } from "@comet/admin";
import { muiGridSortToGql } from "@comet/admin";
import { StackLink } from "@comet/admin";
import { FillSpace } from "@comet/admin";
import { Tooltip } from "@comet/admin";
import { useBufferedRowCount } from "@comet/admin";
import { useDataGridExcelExport } from "@comet/admin";
import { useDataGridRemote } from "@comet/admin";
import { usePersistentColumnState } from "@comet/admin";
import { BlockPreviewContent } from "@comet/cms-admin";
import { useContentScope } from "@comet/cms-admin";
import { Alert } from "@mui/material";
import { Box } from "@mui/material";
import { IconButton } from "@mui/material";
import { Typography } from "@mui/material";
import { useTheme } from "@mui/material";
import { Menu } from "@mui/material";
import { MenuItem } from "@mui/material";
import { ListItemIcon } from "@mui/material";
import { ListItemText } from "@mui/material";
import { CircularProgress } from "@mui/material";
import { DataGridPro } from "@mui/x-data-grid-pro";
import { GridLogicOperator } from "@mui/x-data-grid-pro";
import { GridRenderCellParams } from "@mui/x-data-grid-pro";
import { GridSlotsComponent } from "@mui/x-data-grid-pro";
import { GridToolbarProps } from "@mui/x-data-grid-pro";
import { GridColumnHeaderTitle } from "@mui/x-data-grid-pro";
import { GridToolbarQuickFilter } from "@mui/x-data-grid-pro";
import { GridRowOrderChangeParams } from "@mui/x-data-grid-pro";
import { Add as AddIcon } from "@comet/admin-icons";
import { Edit as EditIcon } from "@comet/admin-icons";
import { Info as InfoIcon } from "@comet/admin-icons";
import { Excel as ExcelIcon } from "@comet/admin-icons";

    const booksFragment = gql\`
        fragment BooksForm on Book {
            id
            title author { birthDate }
        }
    \`;

    const booksQuery = gql\`
        query BooksGrid($offset: Int!, $limit: Int!, $sort: [BookSort!], $filter: BookFilter) {
    books(offset: $offset, limit: $limit, sort: $sort, filter: $filter) {
                nodes {
                    ...BooksForm
                }
                totalCount
            }
        }
        \${booksFragment}
    \`;

    

    const deleteBookMutation = gql\`
                mutation DeleteBook($id: ID!) {
                    deleteBook(id: $id)
                }
            \`;


    interface BooksGridToolbarToolbarProps extends GridToolbarProps {
         exportApi: ExportApi;
}
    function BooksGridToolbar({
        exportApi
    }: BooksGridToolbarToolbarProps) {
        return (
            <DataGridToolbar>
                <GridFilterButton />
                <FillSpace />
        <CrudMoreActionsMenu
        slotProps={{
            button: {
                responsive: true
            }
        }}
        overallActions={[
            {
                        label: <FormattedMessage {...messages.downloadAsExcel} />,
                        icon: exportApi.loading ? <CircularProgress size={20} /> : <ExcelIcon />,
                        onClick: () => exportApi.exportGrid(),
                        disabled: exportApi.loading,
                    }
        ]}
    />
        <Button responsive startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add">
        <FormattedMessage id="book.booksForm.newEntry" defaultMessage={\`New Book\`}  />
    </Button>
            </DataGridToolbar>
        );
    }

    

    export function BooksGrid() {
        const client = useApolloClient();
        const intl = useIntl();
        const dataGridProps = { ...useDataGridRemote(), ...usePersistentColumnState("BooksGrid") };
        
        

        

        const columns: GridColDef<GQLBooksFormFragment>[] = [
            {field: "title",
headerName: intl.formatMessage({ id: "book.title", defaultMessage: "Title" }),
flex: 1,
minWidth: 150,},
{...dataGridDateColumn,field: "author_birthDate",
headerName: intl.formatMessage({ id: "book.author.birthDate", defaultMessage: "Author Birthdate" }),
filterable: false,
sortable: false,
valueGetter: (params, row) => row.author?.birthDate && new Date(row.author?.birthDate),
flex: 1,
minWidth: 150,},
                {field: "actions",
headerName: "",
sortable: false,
filterable: false,
type: "actions",
align: "right",
pinned: "right",
width: 84,
disableExport: true,
renderCell: (params) => {
                            return (
                                <>
                                
                                        <IconButton color="primary" component={StackLink} pageName="edit" payload={params.row.id}>
                                            <EditIcon />
                                        </IconButton>
                                        <CrudContextMenu
                                            
                                            onDelete={async () => {
                                                await client.mutate<GQLDeleteBookMutation, GQLDeleteBookMutationVariables>({
                                                    mutation: deleteBookMutation,
                                                    variables: { id: params.row.id },
                                                });
                                            }}
                                            
                                            refetchQueries={[booksQuery]}
                                        />
                                    
                                </>
                            );
                                },}
        ];

        const { filter: gqlFilter,  } = muiGridFilterToGql(columns, dataGridProps.filterModel);

        const { data, loading, error } = useQuery<GQLBooksGridQuery, GQLBooksGridQueryVariables>(booksQuery, {
            variables: {
                filter: gqlFilter, offset: dataGridProps.paginationModel.page * dataGridProps.paginationModel.pageSize, limit: dataGridProps.paginationModel.pageSize, sort: muiGridSortToGql(dataGridProps.sortModel, columns)
            },
        });
        const rowCount = useBufferedRowCount(data?.books.totalCount);
        if (error) throw error;
        const rows = data?.books.nodes ?? [];

        const exportApi = useDataGridExcelExport<GQLBooksGridQuery["books"]["nodes"][0], GQLBooksGridQuery, Omit<GQLBooksGridQueryVariables, "offset" | "limit">>({
        columns,
        variables: {
            ...muiGridFilterToGql(columns, dataGridProps.filterModel)
        },
        query: booksQuery,
        resolveQueryNodes: (data) => data.books.nodes,
        totalCount: data?.books.totalCount ?? 0,
        exportOptions: {
            fileName: "Books",
        },
    });

        return (
            <DataGridPro
                {...dataGridProps}
                rows={rows}
                rowCount={rowCount}
                columns={columns}
                loading={loading}
                slots={{
                                toolbar: BooksGridToolbar as GridSlotsComponent["toolbar"],
                            }}
                            slotProps={{
        toolbar: {                exportApi,    } as BooksGridToolbarToolbarProps  
    }}
                
            />
        );
    }
    "
`;
